File: FirstMaxFirstMin.cpp
#include<iostream>
using namespace std;

int *firstMax(int *arr, int len, int a) {
  if (a >= len) {
    return arr;
  }

  int last = arr[len - 1];
  for (int i = len - 2; i >= a; i--) {
    arr[i+1] = arr[i];
  }
  arr[a] = last;
  return firstMax(arr, len, a+2);
}

void printArray(int *arr, int len) {
  for (int i = 0; i < len; i++) {
    cout << arr[i];
    if (i != len - 1) {
      cout << " ";
    }
  }
  cout << endl;
}

int main(int argc, char *argv[]) {
  int arr[10] = { 5, 6, 8, 43, 45, 52, 57, 57, 90, 91 },
      len = 10;

  printArray(arr, len);

  arr = firstMax(arr, 0, len - 1);

  printArray(arr, len);

  return 0;
}

File: first_max_first_min.rb
#!/usr/bin/ruby
#

def first_max(arr, a)
  len = arr.length

  return arr if a >= len

  last = arr[len - 1]
  i = len - 2
  while i >= a
    arr[i+1] = arr[i]
    i -= 1
  end
  arr[a] = last;

  first_max(arr, a+2)
end

arr = []
9.times { arr << rand(100) }
arr.sort!

puts arr.inspect
puts first_max(arr, 0).inspect
